Apply network gateway rules to iptables

kursatkobya opened the pull request:
- iptables binary used with configured parameters to create rules
- parsing mechanism is changed due to new prerequisites

kursatkobya committed:
Apply network gateway rules to iptables

breakreturn commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -25,9 +25,9 @@
-    m_interfaceInitialized(false)
+Gateway(ID),
+m_internetAccess(false),
+m_interfaceInitialized(false)
```
I've seen us starting to use different styles w.r.t. this. I think we should keep the original indentation. But we should also have an example in the style guides.

kursatkobya replied:
I did not see formatting constructor initializer lists in our style guide. But adding some rule and applying is quite sensible. 

In this case i think indentation should be kept. So i will change it.

kursatkobya replied:
Done

breakreturn commented on libsoftwarecontainer/src/gateway/networkgateway.h:
```diff
@@ -26,6 +26,65 @@
+/**
+ * @brief A rules entry for the treatment of packets.
+ */
+struct IPTableEntry
```
Why a struct? Why not a class with methods that expose members rather than having a struct that exposes the variables directly?

kursatkobya replied:
The struct is designed for use of NetworkGateway class. Before adding mentioned feature this entry struct was inside of the NetworkGateway class. But i thought it would be better if i collect rule related variables/structure band together. Since the IPTableEntry will hold plain data and all its members ( maybe except insertRule function) are needed to be reachable by NetworkGateway class; thus i believe struct was better option. But it would be better if i move  insertRule funtion to private  members. 

Also i can change this structure to a class if it is believed better way. In this case, a question arises, shall we covert all structures inside IPTableEntry to class as well or not ?

kursatkobya replied:
Maybe we can add rule to style guide related this (struct vs. class). It would be quite useful.

kursatkobya replied:
Converted to class

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -102,25 +103,26 @@ ReturnCode NetworkGateway::readConfigElement(const json_t *element)
     // --- TEMPORARY WORKAROUND ---
     // in the wait of activate() being rewritten
-    if (e.defaultTarget == Target::ACCEPT) {
+    if ("ACCEPT" == e.m_defaultTarget) {
```
Why not keep the enum comparison here?

kursatkobya replied:
Holding targets as enum creates a burden like first it should be parsed from string to enum, then wehen it is needed by iptables it should be converted from enum to string. By storing default target as string saves us from this overhead.

tobsan replied:
But we only need to convert them to string in the end when calling iptables. I think that could be worth the effort.

kursatkobya replied:
Done

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -142,71 +144,52 @@ ReturnCode NetworkGateway::parseRule(const json_t *element, std::vector<Rule> &r
-            int port = json_integer_value(element);
-            ports.push_back(port);
+            int port = json_integer_value(val);
+            portList = portList + std::to_string(port) + ",";
```
This looks a bit hack-ish.

kursatkobya replied:
Since it look quite easy to read and also there is + operator programmed for concatenation, it sounds reasonable to me. append method can be used as well but i prefer this one unless there is no objection.

tobsan replied:
Perhaps this is the neatest way to do it since it will be passed to iptables later on.

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -230,6 +213,9 @@ bool NetworkGateway::activateGateway()
 
     if (m_internetAccess) {
         generateIP();
+        for (auto entry:m_entries) {
```
Space please

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -347,3 +333,54 @@ bool NetworkGateway::isBridgeAvailable()
+    } else if ("FORWARD" == m_type) {
+        return "FORWARD";
+    }
+    return "";
```
This is not a very nice solution. An enum would serve the purpose better imo, and could be compared with in insertRule() also.

kursatkobya replied:
I agree this is not good solution. Imho m_type should be already compatible with iptables chai. So i suggest to change configuration value from 
INCOMING to INPUT
OUTGOING to OUTPUT. 

If it is done so, this function wouldn't be even needed.

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -347,3 +333,54 @@ bool NetworkGateway::isBridgeAvailable()
+    iptableCommand = iptableCommand + " -j " + rule.target;
+    log_debug() << "Add network rule : " <<  iptableCommand;
+
+    system(iptableCommand.c_str());
```
Check the return code here. Also, I think we should try to use the glibmm functions available instead of system(), such as spawn_sync()

https://developer.gnome.org/glibmm/stable/group__Spawn.html

tobsan reviewed (changes requested):
General comment: consider using ReturnCode instead of bool.

kursatkobya committed:
Make network configuration and iptables compatible
Store target in enumeration
Use spawn_command_line_sync for running shell command

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -230,6 +221,11 @@ bool NetworkGateway::activateGateway()
     if (m_internetAccess) {
         generateIP();
+
+        for (auto entry:m_entries) {
```
Spaces here

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -347,3 +343,68 @@ bool NetworkGateway::isBridgeAvailable()
 
+ReturnCode IPTableEntry::applyRules()
+{
+    for (auto rule:m_rules) {
```
Spaces here

tobsan commented on libsoftwarecontainer/src/gateway/networkgateway.cpp:
```diff
@@ -92,8 +97,8 @@ ReturnCode NetworkGateway::readConfigElement(const json_t *element)
 
-    e.defaultTarget = parseTarget(readTarget);
-    if (e.defaultTarget == Target::INVALID_TARGET) {
+    e.m_defaultTarget= parseTarget(readTarget);
```
Add a space before the =

kursatkobya committed:
Seperate class for iptables

kursatkobya committed:
Set default policy before adding rules

DunderRoffe commented on libsoftwarecontainer/src/gateway/iptableentry.h:
```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#ifndef IPTABLEENTRY_H_
+#define IPTABLEENTRY_H_
```
Is not this the same as `#pragma once` right?

breakreturn replied:
Yes, we prefer "pragma once" to guard for this

tobsan commented on libsoftwarecontainer/softwarecontainer.conf.in:
```diff
@@ -6,6 +6,7 @@ lxc.pts = 1
 
 @NETWORK_LXC_CONF@
 
+lxc.mount.entry = /sbin sbin none ro,bind 0 0
```
I think this needs to be discussed more (later). /sbin contains all kinds of dangerous commands. Perhaps we could do a check in the LXC template instead to just copy iptables binary or something.

tobsan commented on libsoftwarecontainer/src/gateway/iptableentry.cpp:
```diff
@@ -0,0 +1,132 @@
+
+ReturnCode IPTableEntry::insertRule(Rule rule)
+{
+    std::string iptableCommand = "/sbin/iptables -A " + m_type;
```
Actually, I think you can make spawn_sync search your PATH variable.
